src/ruby/lib/grpc/generic/active_call.rb

```diff
@@ -110,6 +110,20 @@ def initialize(call, marshal, unmarshal, deadline, started: true,
       @metadata_received = metadata_received
       @metadata_sent = started
       @op_notifier = nil
+
+      fail(ArgumentError, 'Already sent md') if started && metadata_to_send
+      @metadata_to_send = metadata_to_send || {} unless started
+      @send_initial_md_mutex = Mutex.new
+    end
+
+    # Sends the initial metadata that has yet to be sent.
+    # Fails if metadata has already been sent for this call.
+    def send_initial_metadata
+      @send_initial_md_mutex.synchronize do
+        fail('Already send initial metadata') if @metadata_sent
```

Comment:
I think we should check the flag and just do nothing if it's already true. Since this can legitimately be called from multiple threads, it's not an error for it to be called more than once.